### PR TITLE
Reduce usage of ByteConverter

### PR DESCRIPTION
**Pull request by quajak**
As discussed in https://github.com/CosmosOS/Cosmos/issues/956 this PR removes the extension methods as the needed Encoding classes and BitConverter are now available.

### CODE REVIEW

File: source/Cosmos.HAL2/BlockDevice/EBR.cs [hunk 1]
```diff
@@ -40,8 +41,8 @@ protected void ParsePartition(byte[] aEBR, UInt32 aLoc)
             }
             else if (xSystemID != 0)
             {
-                UInt32 xStartSector = aEBR.ToUInt32(aLoc + 8);
-                UInt32 xSectorCount = aEBR.ToUInt32(aLoc + 12);
+                UInt32 xStartSector = (uint)BitConverter.ToInt32(aEBR, (int)aLoc + 8);
```
Comment (jp2masa) [inline source/Cosmos.HAL2/BlockDevice/EBR.cs]: It should be `ToUInt32` here.

File: source/Cosmos.HAL2/BlockDevice/EBR.cs [hunk 2]
```diff
@@ -40,8 +41,8 @@ protected void ParsePartition(byte[] aEBR, UInt32 aLoc)
...
+                UInt32 xSectorCount = (uint)BitConverter.ToInt32(aEBR, (int)aLoc + 12);
```
Comment (jp2masa) [inline source/Cosmos.HAL2/BlockDevice/EBR.cs]: `ToUInt32`.

File: source/Cosmos.HAL2/BlockDevice/MBR.cs [hunk 3]
```diff
@@ -4,56 +4,63 @@
...
+            byte xSystemID = aMBR[aLoc + 4];
+            // SystemID = 0 means no partition
+
+            if (xSystemID == 0x5 || xSystemID == 0xF || xSystemID == 0x85)
+            {
+                //Extended Partition Detected
+                //DOS only knows about 05, Windows 95 introduced 0F, Linux introduced 85
+                //Search for logical volumes
+                //http://thestarman.pcministry.com/asm/mbr/PartTables2.htm
+                EBRLocation = (uint)BitConverter.ToInt32(aMBR, (int)aLoc + 8);
```
Comment (jp2masa) [inline source/Cosmos.HAL2/BlockDevice/MBR.cs]: `ToUInt32`.

File: source/Cosmos.HAL2/BlockDevice/MBR.cs [hunk 4]
```diff
@@ -4,56 +4,63 @@
...
+            }
+            else if (xSystemID != 0)
+            {
+                UInt32 xStartSector = (uint)BitConverter.ToInt32(aMBR, (int)aLoc + 8);
```
Comment (jp2masa) [inline source/Cosmos.HAL2/BlockDevice/MBR.cs]: `ToUInt32`.

File: source/Cosmos.HAL2/BlockDevice/MBR.cs [hunk 5]
```diff
@@ -4,56 +4,63 @@
...
+                UInt32 xSectorCount = (uint)BitConverter.ToInt32(aMBR, (int)aLoc + 12);
```
Comment (jp2masa) [inline source/Cosmos.HAL2/BlockDevice/MBR.cs]: `ToUInt32`.

File: source/Cosmos.System2/FileSystem/FAT/FatFileSystem.cs [hunk 6]
```diff
@@ -218,7 +219,7 @@ private void GetFatEntry(uint aEntryNumber, out uint aValue)
                         // We now access the FAT entry as a WORD just as we do for FAT16, but if the cluster number is
                         // EVEN, we only want the low 12-bits of the 16-bits we fetch. If the cluster number is ODD
                         // we want the high 12-bits of the 16-bits we fetch.
-                        uint xResult = xData.ToUInt16(xEntryOffset);
+                        uint xResult = BitConverter.ToUInt32(xData, (int)xEntryOffset);
```
Comment (jp2masa) [inline source/Cosmos.System2/FileSystem/FAT/FatFileSystem.cs]: `ToUInt16`.

File: source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs [hunk 7]
```diff
@@ -368,19 +369,19 @@ public List<FatDirectoryEntry> ReadDirectoryContents(bool aReturnShortFilenames
...
                         // Also save buffer and only copy name if a end Ord marker is found.
-                        string xLongPart = xData.GetUtf16String(i + 1, 5);
+                        string xLongPart = Encoding.Unicode.GetString(xData, (int)i + 1, 5);
 
                         // We have to check the length because 0xFFFF is a valid Unicode codepoint.
                         // So we only want to stop if the 0xFFFF is AFTER a 0x0000. We can determin
                         // this by also looking at the length. Since we short circuit the or, the length
                         // is rarely evaluated.
-                        if (xData.ToUInt16(i + 14) != 0xFFFF || xLongPart.Length == 5)
+                        if (BitConverter.ToInt32(xData, (int)i + 14) != 0xFFFF || xLongPart.Length == 5)
```
Comment (jp2masa) [inline source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs]: `ToUInt16`.

File: source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs [hunk 8]
```diff
@@ -368,19 +369,19 @@ public List<FatDirectoryEntry> ReadDirectoryContents(bool aReturnShortFilenames
...
                         {
-                            xLongPart = xLongPart + xData.GetUtf16String(i + 14, 6);
+                            xLongPart = xLongPart + Encoding.Unicode.GetString(xData, (int)i + 14, 6);
 
-                            if (xData.ToUInt16(i + 28) != 0xFFFF || xLongPart.Length == 11)
+                            if (BitConverter.ToInt32(xData, (int)i + 28) != 0xFFFF || xLongPart.Length == 11)
```
Comment (jp2masa) [inline source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs]: `ToUInt16`.

File: source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs [hunk 9]
```diff
@@ -458,16 +460,16 @@ public List<FatDirectoryEntry> ReadDirectoryContents(bool aReturnShortFilenames
...
                                     else
                                     {
-                                        xName = xData.GetAsciiString(i, 11).TrimEnd();
+                                        xName = Encoding.ASCII.GetString(xData, (int)i, 11).TrimEnd();
                                     }
                                 }
                             }
 
-                            uint xFirstCluster = (uint)(xData.ToUInt16(i + 20) << 16 | xData.ToUInt16(i + 26));
+                            uint xFirstCluster = ((uint)BitConverter.ToInt32(xData, (int)i + 20) << 16 | (uint)BitConverter.ToInt32(xData, (int)i + 26));
```
Comment (jp2masa) [inline source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs]: `ToUInt16` (both).

File: source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs [hunk 10]
```diff
@@ -458,16 +460,16 @@ public List<FatDirectoryEntry> ReadDirectoryContents(bool aReturnShortFilenames
...
 
                             if (xTest == 0)
                             {
-                                uint xSize = xData.ToUInt32(i + 28);
+                                uint xSize = (uint)BitConverter.ToInt32(xData, (int)i + 28);
```
Comment (jp2masa) [inline source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs]: `ToUInt32`.

File: source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs [hunk 11]
```diff
@@ -482,7 +484,7 @@ public List<FatDirectoryEntry> ReadDirectoryContents(bool aReturnShortFilenames
                             else if (xTest == FatDirectoryEntryAttributeConsts.Directory)
                             {
                                 string xFullPath = Path.Combine(mFullPath, xName);
-                                uint xSize = xData.ToUInt32(i + 28);
+                                uint xSize = (uint)BitConverter.ToInt32(xData, (int)i + 28);
```
Comment (jp2masa) [inline source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs]: `ToUInt32`.

File: Tests/Cosmos.Kernel.Tests.Fat/System.IO/FileTest.cs [hunk 12]
```diff
@@ -154,16 +154,16 @@ public static void Execute(Debugger mDebugger)
...
                 xFile2.Read(xReadBuff, 0, xWriteBuff.Length);
-                mDebugger.Send("xWriteBuff=" + xWriteBuff.GetUtf8String(0, (uint)xWriteBuff.Length));
-                mDebugger.Send("xReadBuff =" + xReadBuff.GetUtf8String(0, (uint)xReadBuff.Length));
-                string xWriteBuffAsString = xWriteBuff.GetUtf8String(0, (uint)xWriteBuff.Length);
-                string xReadBuffAsString = xReadBuff.GetUtf8String(0, (uint)xReadBuff.Length);
+                mDebugger.Send("xWriteBuff=" + Encoding.UTF8.GetString(xWriteBuff));
+                mDebugger.Send("xReadBuff =" + Encoding.UTF8.GetString(xReadBuff));
+                string xWriteBuffAsString = Encoding.UTF8.GetString(xWriteBuff);
+                string xReadBuffAsString = Encoding.UTF8.GetString(xReadBuff);
                 mDebugger.Send("xWriteBuffAsString=" + xWriteBuffAsString);
```
Comment (fanoI) [inline Tests/Cosmos.Kernel.Tests.Fat/System.IO/FileTest.cs]: As Encoding.UTF8.GetString() is a costly operation do it only once the code could be changed in this way:

```cs
string xWriteBuffAsString = Encoding.UTF8.GetString(xWriteBuff);
string xReadBuffAsString = Encoding.UTF8.GetString(xReadBuff);
mDebugger.Send("xWriteBuff=" + xWriteBuffAsString ));
mDebugger.Send("xReadBuff =" + xReadBuffAsString );
```

File: source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs [hunk 13]
```diff
@@ -787,7 +789,7 @@ internal void SetLongFilenameEntryMetadataValue(uint aEntryHeaderDataOffset, Fat
             if (xData.Length > 0)
             {
                 var xValue = new byte[aEntryMetadata.DataLength];
-                xValue = aValue.GetUtf16Bytes(0, aEntryMetadata.DataLength / 2);
+                xValue = Encoding.Unicode.GetBytes(aValue.ToCharArray(), 0, (int)aEntryMetadata.DataLength / 2);
```
Comment (fanoI) [inline source/Cosmos.System2/FileSystem/FAT/Listing/FatDiretoryEntry.cs]: As ``aValue`` is a C# string already the conversion using ToCharArray() is not needed, as the last parameter is charCount I think the division per 2 is not needed anymore.

File: source/Cosmos.HAL2/BlockDevice/EBR.cs (hunk 1 above)
Comment (quajak) [reply source/Cosmos.HAL2/BlockDevice/EBR.cs]: ToUInt32 is not working. I am getting an Arg_InvalidOperationException when trying to use it. I think it has to be plugged.

**Comment by quajak** [comment]
I have also added the BitConverter.ToUintXX methods.

**Comment by fanoI** [comment]
Excellent!
I see it continues to fail but at least this a managed exception and not a stack corruption:

> Exception occurred: Exception: ArgumentOutOfRange_IndexCountBuffer

when it tries to change the Volume Label to "TEST" apparently sadly there is no stack trace support but you can follow from here adding debug logs from where this exception comes.

**Pull request merged**